### PR TITLE
Add HTTP forwarder.

### PR DESCRIPTION
**Pull request by kacf**
Ticket: MEN-6663

### CODE REVIEW

File: mender-auth/http_forwarder/http_forwarder.cpp
Comment (lluiscampos) [inline mender-auth/http_forwarder/http_forwarder.cpp:259]: This file was so easy to read. Amazing.

**Review by lluiscampos — COMMENTED** [review]
I looks promising. I'll wait for the final review once you submit the fix you described today in SU :+1:

File: mender-auth/CMakeLists.txt
Comment (kacf) [reply mender-auth/CMakeLists.txt:20]: @lluiscampos: I think that's best, yes. It's not really a core part of the HTTP library, more of an extension, just like this one.

**Comment by kacf** [comment]
Updated! @lluiscampos: The only real difference from the original you saw is the addition of the "[chore(http): Standardize cancellation handling on one variable.](https://github.com/mendersoftware/mender/pull/1390/commits/2fa34d9a96f7d6d849af3f12314ddf7a566ddec7)" commit, which was required in order to deal with the nasty asan issue.

File: common/http.hpp
Comment (lluiscampos) [inline common/http.hpp:286]: Looking at the resulting code, I think it would make more sense to call the flag `ongoing_` or `active_` and use the opposite logic: initialize to false when creating the object, set to true when operation starts, set to false again on cancel or termination. At this point it seems like we used `cancelled_` "for historical reasons".

It is an observation, I am not saying that we should change it.

File: common/http/platform/beast/http.cpp
Comment (kacf) [reply common/http/platform/beast/http.cpp:193]: After a reference has been assigned, it behaves as if it were the original object, so using a reference here does not automatically trigger the use of a reference in the lambda. For that we would have to use the `[&cancelled]` expression. So in other words, there is another copy being made when we capture the variable in the lambda, and using a reference here is just a micro-optimization to avoid making two copies.

File: common/http.hpp
Comment (kacf) [reply common/http.hpp:286]: Yes, I get your point.

But I agree I don't think we should change it now, since the `cancelled_` method is used in a lot of other places too, so at least we're being consistent. Changing it everywhere would be a lot of work, since it's not only a rename, but a logic inversion.

File: common/http/platform/beast/http.cpp
Comment (lluiscampos) [reply common/http/platform/beast/http.cpp:193]: Aha! Thank you.

File: mender-auth/http_forwarder/http_forwarder_test.cpp
Comment (lluiscampos) [inline mender-auth/http_forwarder/http_forwarder_test.cpp:455]: As this test is explicitly terminating, we don't need to oversize the header, do we?

**Review by lluiscampos — APPROVED** [review]
I finalized now the review. It looks very good!

File: mender-auth/http_forwarder/http_forwarder.cpp
Comment (oleorhagen) [inline mender-auth/http_forwarder/http_forwarder.cpp:17]: Nit: Include <string>

File: common/http/platform/beast/http.cpp
Comment (oleorhagen) [reply common/http/platform/beast/http.cpp:193]: C++ at its finest :laughing:

**Review by oleorhagen — APPROVED** [review]
Looks good imo !

Also have a look at all the consts I was able to add!

```diff --git a/mender-auth/http_forwarder.hpp b/mender-auth/http_forwarder.hpp
index 57dcd9d7..cb57c547 100644
--- a/mender-auth/http_forwarder.hpp
+++ b/mender-auth/http_forwarder.hpp
@@ -71,12 +71,12 @@ public:
 	string GetUrl() const;
 
 private:
-	void RequestHeaderHandler(http::ExpectedIncomingRequestPtr exp_req);
-	void RequestBodyHandler(http::IncomingRequestPtr req, error::Error err);
+	void RequestHeaderHandler(const http::ExpectedIncomingRequestPtr exp_req);
+	void RequestBodyHandler(const http::IncomingRequestPtr req, const error::Error err);
 	void ResponseHeaderHandler(
-		http::IncomingRequestPtr req_in, http::ExpectedIncomingResponsePtr exp_resp_in);
+		const http::IncomingRequestPtr req_in, const http::ExpectedIncomingResponsePtr exp_resp_in);
 	void ResponseBodyHandler(
-		http::IncomingRequestPtr req_in, http::ExpectedIncomingResponsePtr exp_resp_in);
+		const http::IncomingRequestPtr req_in, const http::ExpectedIncomingResponsePtr exp_resp_in);
 
 	log::Logger logger_;
 	events::EventLoop &event_loop_;
diff --git a/mender-auth/http_forwarder/CMakeLists.txt b/mender-auth/http_forwarder/CMakeLists.txt
index 03ca78fe..4b466f35 100644
--- a/mender-auth/http_forwarder/CMakeLists.txt
+++ b/mender-auth/http_forwarder/CMakeLists.txt
@@ -6,6 +6,11 @@ target_link_libraries(mender_http_forwarder PUBLIC
 )
 
 add_executable(http_forwarder_test EXCLUDE_FROM_ALL http_forwarder_test.cpp)
-target_link_libraries(http_forwarder_test PUBLIC mender_http_forwarder common_testing main_test gmock)
+target_link_libraries(http_forwarder_test PUBLIC
+  mender_http_forwarder
+  common_testing
+  main_test
+  gmock
+)
 gtest_discover_tests(http_forwarder_test)
 add_dependencies(tests http_forwarder_test)
diff --git a/mender-auth/http_forwarder/http_forwarder.cpp b/mender-auth/http_forwarder/http_forwarder.cpp
index 3f052586..22ea544d 100644
--- a/mender-auth/http_forwarder/http_forwarder.cpp
+++ b/mender-auth/http_forwarder/http_forwarder.cpp
@@ -15,6 +15,7 @@
 #include <mender-auth/http_forwarder.hpp>
 
 #include <random>
+#include <string>
 
 namespace mender {
 namespace auth {
@@ -96,31 +97,31 @@ string Server::GetUrl() const {
 	return server_.GetUrl();
 }
 
-void Server::RequestHeaderHandler(http::ExpectedIncomingRequestPtr exp_req) {
+void Server::RequestHeaderHandler(const http::ExpectedIncomingRequestPtr exp_req) {
 	if (!exp_req) {
 		logger_.Error("Error in incoming request: " + exp_req.error().String());
 		return;
 	}
-	auto &req_in = exp_req.value();
+	const auto &req_in = exp_req.value();
 
 	ForwardObjectPtr connection {new ForwardObject(client_config_, event_loop_)};
 	connections_[req_in] = connection;
 	connection->logger_ = logger_.WithFields(log::LogField {"request", req_in->GetPath()});
 	connection->req_in_ = req_in;
 
-	auto final_url = http::JoinUrl(target_url_, req_in->GetPath());
-	auto req_out = make_shared<http::OutgoingRequest>();
+	const auto final_url = http::JoinUrl(target_url_, req_in->GetPath());
+	const auto req_out = make_shared<http::OutgoingRequest>();
 	req_out->SetMethod(req_in->GetMethod());
 	req_out->SetAddress(final_url);
-	for (auto header : req_in->GetHeaders()) {
+	for (const auto &header : req_in->GetHeaders()) {
 		req_out->SetHeader(header.first, header.second);
 	}
 	connection->req_out_ = req_out;
 
-	auto exp_body_reader = req_in->MakeBodyAsyncReader();
+	const auto exp_body_reader = req_in->MakeBodyAsyncReader();
 	if (exp_body_reader) {
-		auto body_reader = exp_body_reader.value();
-		auto generated = make_shared<bool>(false);
+		const auto body_reader = exp_body_reader.value();
+		const auto generated = make_shared<bool>(false);
 		req_out->SetAsyncBodyGenerator([body_reader, generated]() -> io::ExpectedAsyncReaderPtr {
 			// We can only do this once, because the incoming request body is not
 			// seekable.
@@ -140,8 +141,8 @@ void Server::RequestHeaderHandler(http::ExpectedIncomingRequestPtr exp_req) {
 		return;
 	} // else: if body is missing we don't need to do anything.
 
-	auto &cancelled = cancelled_;
-	auto err = connection->client_.AsyncCall(
+	const auto &cancelled = cancelled_;
+	const auto err = connection->client_.AsyncCall(
 		req_out,
 		[cancelled, this, req_in](http::ExpectedIncomingResponsePtr exp_resp) {
 			if (!*cancelled) {
@@ -155,13 +156,13 @@ void Server::RequestHeaderHandler(http::ExpectedIncomingRequestPtr exp_req) {
 		});
 }
 
-void Server::RequestBodyHandler(http::IncomingRequestPtr req_in, error::Error err) {
-	auto maybe_connection = connections_.find(req_in);
+void Server::RequestBodyHandler(const http::IncomingRequestPtr req_in, const error::Error err) {
+	const auto maybe_connection = connections_.find(req_in);
 	if (maybe_connection == connections_.end()) {
 		// Can happen if the request was cancelled.
 		return;
 	}
-	auto &connection = maybe_connection->second;
+	const auto &connection = maybe_connection->second;
 
 	if (err != error::NoError) {
 		connection->logger_.Error("Error while reading incoming request body: " + err.String());
@@ -169,7 +170,7 @@ void Server::RequestBodyHandler(http::IncomingRequestPtr req_in, error::Error er
 		return;
 	}
 
-	auto exp_resp_out = connection->req_in_->MakeResponse();
+	const auto exp_resp_out = connection->req_in_->MakeResponse();
 	if (!exp_resp_out) {
 		connection->logger_.Error(
 			"Could not make outgoing response: " + exp_resp_out.error().String());
@@ -180,8 +181,8 @@ void Server::RequestBodyHandler(http::IncomingRequestPtr req_in, error::Error er
 }
 
 void Server::ResponseHeaderHandler(
-	http::IncomingRequestPtr req_in, http::ExpectedIncomingResponsePtr exp_resp_in) {
-	auto connection = connections_[req_in];
+	const http::IncomingRequestPtr req_in, const http::ExpectedIncomingResponsePtr exp_resp_in) {
+	const auto connection = connections_.at(req_in);
 
 	if (!exp_resp_in) {
 		connection->logger_.Error("Error in incoming response: " + exp_resp_in.error().String());
@@ -189,16 +190,16 @@ void Server::ResponseHeaderHandler(
 		return;
 	}
 	connection->resp_in_ = exp_resp_in.value();
-	auto &resp_in = connection->resp_in_;
+	const auto &resp_in = connection->resp_in_;
 
-	auto &resp_out = connection->resp_out_;
+	const auto &resp_out = connection->resp_out_;
 
 	resp_out->SetStatusCodeAndMessage(resp_in->GetStatusCode(), resp_in->GetStatusMessage());
-	for (auto header : resp_in->GetHeaders()) {
+	for (const auto &header : resp_in->GetHeaders()) {
 		resp_out->SetHeader(header.first, header.second);
 	}
 
-	auto exp_body_reader = resp_in->MakeBodyAsyncReader();
+	const auto exp_body_reader = resp_in->MakeBodyAsyncReader();
 	if (exp_body_reader) {
 		resp_out->SetAsyncBodyReader(exp_body_reader.value());
 	} else if (exp_body_reader.error().code != http::MakeError(http::BodyMissingError, "").code) {
@@ -208,20 +209,20 @@ void Server::ResponseHeaderHandler(
 		return;
 	} // else: if body is missing we don't need to do anything.
 
-	auto &cancelled = cancelled_;
+	const auto &cancelled = cancelled_;
 	auto err = resp_out->AsyncReply([cancelled, this, req_in](error::Error err) {
 		if (*cancelled) {
 			return;
 		}
 
 		if (err != error::NoError) {
-			connections_[req_in]->logger_.Error(
+			connections_.at(req_in)->logger_.Error(
 				"Error while forwarding response to client: " + err.String());
 			connections_.erase(req_in);
 			return;
 		}
 
-		auto &connection = connections_[req_in];
+		const auto &connection = connections_.at(req_in);
 		connection->incoming_request_finished = true;
 		if (connection->outgoing_request_finished) {
 			// We are done, remove connection.
@@ -236,8 +237,8 @@ void Server::ResponseHeaderHandler(
 }
 
 void Server::ResponseBodyHandler(
-	http::IncomingRequestPtr req_in, http::ExpectedIncomingResponsePtr exp_resp_in) {
-	auto &connection = connections_[req_in];
+	const http::IncomingRequestPtr req_in, const http::ExpectedIncomingResponsePtr exp_resp_in) {
+	const auto &connection = connections_.at(req_in);
 
 	if (!exp_resp_in) {
 		connection->logger_.Error(
```

Joking a little bit obviously, but the interface we could const I think

File: mender-auth/http_forwarder/http_forwarder_test.cpp
Comment (kacf) [reply mender-auth/http_forwarder/http_forwarder_test.cpp:455]: Sort of true, but if we cancel exactly at the body length, then there is a race condition. The request is a complete request then and the server may actually have time to reply before the cancel goes through. Unlikely, since it's going to write a whole MB, but I'd rather have it unambiguous.

**Comment by kacf** [comment]
> Joking a little bit obviously, but the interface we could const I think

Yes, and as you know I disagree that they are needed in many places where they can technically be used (because they prevent variable reuse, and do not provide value, IMHO).

However, in your suggestion you added some consts that I would consider harmful. Take for example `const http::IncomingRequestPtr req_in`. It's not the request which is made const, it's the pointer. Which means that no real modifications are being prevented, however, move optimizations *are* being prevented, because moving is not allowed on const objects, even if they are about to go out of scope.

So I stick to my previous rule, which I think is also easier to follow: Const on references, non-const on values, with occasional const on the latter if you really want to underline its constness.

File: common/http_test_helpers.hpp
Comment (kacf) [reply common/http_test_helpers.hpp:3]: Which `Server`..? Not sure I understand...

File: mender-auth/http_forwarder/http_forwarder.cpp
Comment (kacf) [reply mender-auth/http_forwarder/http_forwarder.cpp:23]: Why? Then I have to go back and forth.

File: mender-auth/http_forwarder/http_forwarder.cpp
Comment (kacf) [reply mender-auth/http_forwarder/http_forwarder.cpp:71]: Because C++ doesn't allow direct capture of member variables. However, as indicated in another comment, I should make this a reference, which avoids one copy of two (still one copy in the lambda, which we want).

File: mender-auth/http_forwarder/http_forwarder_test.cpp
Comment (kacf) [reply mender-auth/http_forwarder/http_forwarder_test.cpp:64]: Sure, why not.

File: mender-auth/http_forwarder/http_forwarder_test.cpp
Comment (kacf) [reply mender-auth/http_forwarder/http_forwarder_test.cpp:256]: Removed.

File: mender-auth/http_forwarder/http_forwarder.cpp
Comment (kacf) [reply mender-auth/http_forwarder/http_forwarder.cpp:212]: I won't argue with readability and familiarity! Fixed.

File: mender-auth/http_forwarder/http_forwarder.cpp
Comment (kacf) [reply mender-auth/http_forwarder/http_forwarder.cpp:17]: Thanks, I ended up not using this.

File: common/http_test_helpers.hpp
Comment (oleorhagen) [reply common/http_test_helpers.hpp:3]: The one from `http` lib

File: mender-auth/http_forwarder/http_forwarder.cpp
Comment (oleorhagen) [reply mender-auth/http_forwarder/http_forwarder.cpp:21]: Hmm, just a matter of preference I guess. I find I like them in the header file, but obviously YMMV

**Comment by oleorhagen** [comment]
> > Joking a little bit obviously, but the interface we could const I think
> 
> Yes, and as you know I disagree that they are needed in many places where they can technically be used (because they prevent variable reuse, and do not provide value, IMHO).
> 
> However, in your suggestion you added some consts that I would consider harmful. Take for example `const http::IncomingRequestPtr req_in`. It's not the request which is made const, it's the pointer. Which means that no real modifications are being prevented, however, move optimizations _are_ being prevented, because moving is not allowed on const objects, even if they are about to go out of scope.
> 
> So I stick to my previous rule, which I think is also easier to follow: Const on references, non-const on values, with occasional const on the latter if you really want to underline its constness.

https://google.github.io/styleguide/cppguide.html#Use_of_const

```
We strongly recommend using const in APIs (i.e., on function parameters, methods, and non-local variables) wherever it is meaningful and accurate. This provides consistent, mostly compiler-verified documentation of what objects an operation can mutate. Having a consistent and reliable way to distinguish reads from writes is critical to writing thread-safe code, and is useful in many other contexts as well. In particular
```

**Comment by kacf** [comment]
Yes, but the keyword there is "meaningful". Const has no effect on the API for non-reference variables.

File: common/http_test_helpers.hpp
Comment (kacf) [reply common/http_test_helpers.hpp:3]: Ah no. The server is being used by the forwarder, so we can't keep it in test files.

**Comment by kacf** [comment]
> Yes, but the keyword there is "meaningful". Const has no effect on the API for non-reference variables.

Take a look at the C++ library for example. I have yet to find even a single non-reference const parameter.

**Review by vpodzime — COMMENTED** [review]
Just a tiny little nitpick.

**Pull request merged**
